In [1]:
import os

#モジュール追加用pathの設定
import sys
#sys.path.append('..')

#モジュールのインポート
import numpy as np

**70.py**

In [15]:
import gensim
import re
import torch
import numpy as np


w2v_path = "GoogleNews-vectors-negative300.bin" #todo GoogleNews-vectors-negative300.binのパスを指定
data_path = "./data/"#todo ファイル出力ディレクトリを指定

w2v_model = gensim.models.KeyedVectors.load_word2vec_format(w2v_path, binary=True) #todo gensimでword2vecのモデルを読み込む

category2num = {"b": 0, "t": 1, "e": 2, "m": 3}

In [40]:
w2v_model['the'].shape

(300,)

In [41]:
import numpy as np

a = np.arange(1,301)
print(a.shape)
print(type(a))
print('-'*20)

b = [0]*10671
print(type(b))
print('-'*20)

c = torch.tensor(b)
print(c.shape)
print(type(c))
print('-'*20)

(300,)
<class 'numpy.ndarray'>
--------------------
<class 'list'>
--------------------
torch.Size([10671])
<class 'torch.Tensor'>
--------------------


In [13]:
a = np.array([[1, 2, 3], [4, 5, 6]])
print(np.sum(a))
print(np.sum(a, axis=0))

21
[5 7 9]


In [24]:
# タイトルを受け取り，単語ベクトルの平均を返す関数
def get_feature(title):
    word_list = title.split(' ')#todo タイトルをスペースで分割
    vec_list = []
    for word in word_list:
        try:
            vec =w2v_model[word] #todo wordを意味するベクトルを取得
#             print('bb')
        except KeyError:
            vec = np.zeros(300)#todo すべての要素が0のベクトルを代入
#             print('aa')
        vec_list.append(vec)    
    vec_np = np.array(vec_list) # numpyのarrayに変換
    feature = np.sum(vec_np, axis=0)/len(vec_np)# 平均ベクトルを計算
    print(feature.shape)
    return feature


def get_data(fname):
    label_list = []
    feature_list = []
    with open(fname, encoding="utf-8") as f:
        for line in f:
            if not line:
                continue
            data = line.split('\t')
            title = data[1]#todo dataからタイトルを取り出す
            feature = get_feature(title) 
            feature_list.append(feature)
            label = category2num[data[0]]
            label_list.append(label)
#     print(len(feature_list))
#     print(label_list[0])
#     print(len(feature_list))
#     print(len(label_list))
#     print(feature_list[1])
#     print(label_list[1])
    features = torch.tensor(feature_list)#todo feature_listをtensorに変換
    labels = torch.tensor(label_list)#todo label_listをtensorに変換
    return features, labels


train_x, train_y = get_data(data_path + "train.txt")
valid_x, valid_y = get_data(data_path + "valid.txt")
test_x, test_y = get_data(data_path + "test.txt")

# 保存
torch.save(train_x, data_path + "train_x.pt")
torch.save(train_y, data_path + "train_y.pt")
torch.save(valid_x, data_path + "valid_x.pt")
torch.save(valid_y, data_path + "valid_y.pt")
torch.save(test_x, data_path + "test_x.pt")
torch.save(test_y, data_path + "test_y.pt")

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)

(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)
(300,)


実行したところ確かに上記のファイルが作成できた。

**77.py**

In [2]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
import time

In [35]:
train_x, train_y = torch.load("./data/train_x.pt"), torch.load("./data/train_y.pt")
valid_x, valid_y = torch.load("./data/valid_x.pt"), torch.load("./data/valid_y.pt")

# for i, val in enumerate(train_x):
#     print(i, val)
print(train_x.shape)
print(train_y.shape)
print(valid_x.shape)
print(valid_y.shape)
# print(train_x[0])
print(valid_x[0].shape)

torch.Size([10671, 300])
torch.Size([10671])
torch.Size([1334, 300])
torch.Size([1334])
torch.Size([300])


In [78]:
loss = nn.CrossEntropyLoss()
input = torch.randn(1, 4, requires_grad=True)
print(input)
target = torch.empty(1, dtype=torch.long).random_(5)
print(target)
output = loss(input, target)
print(output)
output.backward()

tensor([[-2.1664,  2.0389,  0.7590, -0.8060]], requires_grad=True)
tensor([2])
tensor(1.5808, grad_fn=<NllLossBackward>)


In [ ]:
a = [1, 3, 5]
b = [0, 3, 4]
()

In [53]:
# 単層ニューラルネットワークを定義
class SingleLayerNN(nn.Module):
    def __init__(self, embedding_dim, num_labels):
        super(SingleLayerNN, self).__init__()
        self.linear = nn.Linear(embedding_dim, num_labels) #todo linear層を追加
    
    def forward(self, x):
        h1 = self.linear(x) #todo xをlinear層に入力
        softmax = nn.Softmax()
        return softmax(h1)#todo h1をsoftmaxしたものを返す

class MyDataset(Dataset):
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __len__(self):
        return len(self.x) #todo データサイズを返す

    def __getitem__(self, idx):
        out_x = self.x[idx] #todo idx番目のxの要素を返す
        out_y = self.y[idx] #todo idx番目のyの要素を返す
        return out_x, out_y
        
torch.manual_seed(1)

embedding_dim = 300
num_labels = 4
NUM_EPOCH = 100

train_x, train_y = torch.load("./data/train_x.pt"), torch.load("./data/train_y.pt")
valid_x, valid_y = torch.load("./data/valid_x.pt"), torch.load("./data/valid_y.pt")

# データとラベルを1つにまとめる
dataset = MyDataset(train_x, train_y)

loss_fn = nn.CrossEntropyLoss()#todo 損失関数としてCrossEntropyLossを用いる


def train(model, train_loader):
    model.train()
    for data, target in train_loader:
        optimizer.zero_grad()
        pred = model.forward(data.type(torch.FloatTensor))#todo dataをmodelに入力
        loss = loss_fn(pred, target)#todo predとtargetと損失関数からロスを計算
#         print(loss)
        loss.backward() # 逆誤差伝搬を実施
        optimizer.step() # パラメータを更新

        
def evaluation(model, data, target):
    model.eval()
    with torch.no_grad():
        pred =  model(data.type(torch.FloatTensor))#todo dataをmodelに入力
#        print(pred.shape) 1334*4
        loss = loss_fn(pred, target)#todo predとtargetと損失関数からロスを計算
        pred_labels = torch.argmax(pred, axis=1)#todo 推定したラベルを代入
#         print(pred_labels.shape, target.shape) #1334, 1334
        acc = (pred_labels == target).sum().item() / len(pred_labels)#todo pred_labelsとtargetから正解率を計算
    return loss.item(), acc


batch_size_list = [1, 2, 4, 8, 16, 32, 64, 128, 256] #todo バッチサイズの値のリスト．1,2,4,8,…が入る

for batch_size in batch_size_list:
    model = SingleLayerNN(embedding_dim, num_labels)
    optimizer = optim.SGD(model.parameters(), lr=0.1)

    # ミニバッチを扱うためのデータローダを作成
    train_loader = DataLoader(dataset, batch_size=batch_size, shuffle=True)

    for epoch in range(NUM_EPOCH):
        start_time = time.time() # 学習開始時の時間を取得
        train(model, train_loader)
        took_time = time.time() - start_time #todo 学習にかかった時間を代入

        valid_loss, valid_acc = evaluation(model, valid_x, valid_y)

        if epoch % 10 == 0:
            print(f"batch size:{batch_size}\tepoch: {epoch}")
            print(f"train time: {took_time}")
            print(f"<valid>\tloss: {valid_loss}\tacc: {valid_acc}")
    
    print("\n================\n")


C:\Users\ken19\anaconda3\lib\site-packages\ipykernel_launcher.py:10: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  # Remove the CWD from sys.path while we load stuff.


batch size:1	epoch: 0
train time: 6.940154314041138
<valid>	loss: 0.9587206244468689	acc: 0.800599700149925
batch size:1	epoch: 10
train time: 6.73320198059082
<valid>	loss: 0.9397221803665161	acc: 0.8013493253373314
batch size:1	epoch: 20
train time: 6.675987482070923
<valid>	loss: 0.9378631114959717	acc: 0.8020989505247377
batch size:1	epoch: 30
train time: 6.556579113006592
<valid>	loss: 0.936996340751648	acc: 0.8035982008995503
batch size:1	epoch: 40
train time: 5.603638648986816
<valid>	loss: 0.9364750981330872	acc: 0.802848575712144
batch size:1	epoch: 50
train time: 5.879449129104614
<valid>	loss: 0.9360454082489014	acc: 0.8020989505247377
batch size:1	epoch: 60
train time: 5.602369785308838
<valid>	loss: 0.8930368423461914	acc: 0.8538230884557722
batch size:1	epoch: 70
train time: 5.779035329818726
<valid>	loss: 0.8867973685264587	acc: 0.8590704647676162
batch size:1	epoch: 80
train time: 5.637355089187622
<valid>	loss: 0.885372519493103	acc: 0.8598200899550225
batch size:1	epo

batch size:128	epoch: 30
train time: 0.3271059989929199
<valid>	loss: 0.996565580368042	acc: 0.7976011994002998
batch size:128	epoch: 40
train time: 0.32499051094055176
<valid>	loss: 0.9856399893760681	acc: 0.7983508245877061
batch size:128	epoch: 50
train time: 0.3119633197784424
<valid>	loss: 0.9785651564598083	acc: 0.7991004497751124
batch size:128	epoch: 60
train time: 0.31694889068603516
<valid>	loss: 0.9735413193702698	acc: 0.7998500749625187
batch size:128	epoch: 70
train time: 0.32466816902160645
<valid>	loss: 0.9697608947753906	acc: 0.7991004497751124
batch size:128	epoch: 80
train time: 0.504683256149292
<valid>	loss: 0.966791570186615	acc: 0.7991004497751124
batch size:128	epoch: 90
train time: 0.4785299301147461
<valid>	loss: 0.9643874168395996	acc: 0.7998500749625187


batch size:256	epoch: 0
train time: 0.45450282096862793
<valid>	loss: 1.3434481620788574	acc: 0.676911544227886
batch size:256	epoch: 10
train time: 0.43000054359436035
<valid>	loss: 1.1388838291168213	acc: 

たしかに実行できた。epoch数が増えるとaccの値が大きくなっていることが確認できた。今回は100だったのでaccの値があまり大きくならなかった。また、batch_size = 1のときは80%から85%に急に飛んでいるが、他のbatch_sizeでは同様の現象は見られなかった。batch_sizeが小さいほどtrain timeが小さくなることも確認できた。